src/interfaces/interfaces.ts

```diff
@@ -54,34 +30,143 @@ namespace interfaces {
+        {
+            factoryType: "toDynamicValue";
+            type: "DynamicValue";
+        }
+
+    export interface ConstructorValueProvider<TActivated>
+        extends ValueProvider<TActivated,TActivated>, Clonable<ConstructorValueProvider<TActivated>>{
+        type: "Constructor";
+    }
+
+    export interface FactoryValueProvider<TActivated> extends
+        FactoryTypeValueProvider<TActivated, (context:interfaces.Context) => TActivated>, Clonable<FactoryValueProvider<TActivated>> {
+        factoryType:"toFactory";
+        type: "Factory"
+    }
+
+    export interface ProviderValueProvider<TActivated> extends
+        FactoryTypeValueProvider<TActivated, (context:interfaces.Context) => TActivated>, Clonable<ProviderValueProvider<TActivated>> {
+        factoryType:"toProvider";
+        type: "Provider";
+    }
+
+    export interface NotConfiguredValueProvider extends ValueProvider<never,never>, Clonable<NotConfiguredValueProvider>{
+        type: "NotConfigured";
+    }
+
+    export type ValueProviderType<TActivated> =
+        ConstantValueProvider<TActivated> |
+        InstanceValueProvider<TActivated> |
+        DynamicValueProvider<TActivated> |
+        ConstructorValueProvider<TActivated> |
+        FactoryValueProvider<TActivated> |
+        ProviderValueProvider<TActivated> |
+        NotConfiguredValueProvider
+
+    export type BindingType = ValueProviderType<unknown>["type"];
+
+    export interface BindingTypeEnum {
+        ConstantValue: ConstantValueProvider<unknown>["type"];
+        Constructor: ConstructorValueProvider<unknown>["type"];
+        DynamicValue: DynamicValueProvider<unknown>["type"];
+        Factory: FactoryValueProvider<unknown>["type"];
+        Instance: InstanceValueProvider<unknown>["type"];
+        Provider: ProviderValueProvider<unknown>["type"];
+        NotConfigured: NotConfiguredValueProvider["type"];
+    }
+
+    export interface Scope<T>{
+        get(binding:Binding<T>,request:Request):Promise<T>|T|undefined
+        set(binding:interfaces.Binding<T>,request:Request,resolved:T|Promise<T>):T | Promise<T>
+    }
+
+    export interface SingletonScope<T> extends Scope<T>, Clonable<SingletonScope<T>>{
+        type:"Singleton",
+        resolved: T | Promise<T> | undefined;
+    }
+
+    export interface TransientScope<T> extends Scope<T>,Clonable<TransientScope<T>>{
+        type:"Transient"
+    }
+
+    export interface RequestResolveScope<T> extends Scope<T>, Clonable<RequestResolveScope<T>>{
+        type:"Request"
+    }
+
+    export interface RootRequestScope<T> extends Scope<T>,Clonable<RootRequestScope<T>>{
+        type:"RootRequest"
+    }
+
+    export interface CustomScope<T> extends Scope<T>,Clonable<CustomScope<T>>{
+        type:"Custom"
+    }
+
+    export interface NotConfiguredScope extends Scope<never>,Clonable<NotConfiguredScope>{
+        type:"NotConfigured",
```

Comment:
NotConfiguredScope could be renamed InvalidScope with type Invalid to adhere to previous BindingType/Enum